samples/fpgaperf_counter/fpgaperf_counter.c

```diff
@@ -0,0 +1,646 @@
+// Copyright(c) 2021, Intel Corporation
+//
+// Redistribution  and  use  in source  and  binary  forms,  with  or  without
+// modification, are permitted provided that the following conditions are met:
+//
+// * Redistributions of  source code  must retain the  above copyright notice,
+//   this list of conditions and the following disclaimer.
+// * Redistributions in binary form must reproduce the above copyright notice,
+//   this list of conditions and the following disclaimer in the documentation
+//   and/or other materials provided with the distribution.
+// * Neither the name  of Intel Corporation  nor the names of its contributors
+//   may be used to  endorse or promote  products derived  from this  software
+//   without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING,  BUT NOT LIMITED TO,  THE
+// IMPLIED WARRANTIES OF  MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED.  IN NO EVENT  SHALL THE COPYRIGHT OWNER  OR CONTRIBUTORS BE
+// LIABLE  FOR  ANY  DIRECT,  INDIRECT,  INCIDENTAL,  SPECIAL,  EXEMPLARY,  OR
+// CONSEQUENTIAL  DAMAGES  (INCLUDING,  BUT  NOT LIMITED  TO,  PROCUREMENT  OF
+// SUBSTITUTE GOODS OR SERVICES;  LOSS OF USE,  DATA, OR PROFITS;  OR BUSINESS
+// INTERRUPTION)  HOWEVER CAUSED  AND ON ANY THEORY  OF LIABILITY,  WHETHER IN
+// CONTRACT,  STRICT LIABILITY,  OR TORT  (INCLUDING NEGLIGENCE  OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,  EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#include "fpgaperf_counter.h"
+
+#ifndef __USE_GNU
+#define __USE_GNU 1
+#endif
+#include <pthread.h>
+
+#include <errno.h>
+#include <glob.h>
+#include <regex.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <string.h>
+#include <unistd.h>
+#include <sys/ioctl.h>
+#include <sys/syscall.h>
+#include <libudev.h>
+#include <linux/perf_event.h>
+
+#include <opae/fpga.h>
+#include <opae/log.h>
+#include <opae/properties.h>
+#include <opae/utils.h>
+#include "opae_int.h"
+
+#define PCI_DEV_ADDRS		"/sys/bus/pci/devices/*%x*:*%x*:*%x*.*%x*/"
+
+#define DFL_PERF_FME		PCI_DEV_ADDRS "fpga_region/region*/dfl-fme.*"
+
+#define DFL_PERF_SYSFS		"/sys/bus/event_source/devices/dfl_fme"
+
+#define PERF_EVENT		"event=(0x[0-9a-fA-F]{2}),"
+
+#define PERF_EVTYPE		"evtype=(0x[0-9a-fA-F]{2}),"
+
+#define PERF_PORTID		"portid=(0x[0-9a-fA-F]{2})"
+
+#define PERF_EVENT_PATTERN	PERF_EVENT PERF_EVTYPE PERF_PORTID
+
+#define PERF_CONFIG_PATTERN	"config:([0-9]{1,})-([0-9]{2,})"
+
+#define PARSE_MATCH_INT(_p, _m, _v, _b)				\
+	do {							\
+		errno = 0;					\
+		_v = strtoul(_p + _m, NULL, _b);		\
+		if (errno) {					\
+			OPAE_MSG("error parsing int");		\
+		}						\
+	} while (0)
+
+/* mutex to protect fpga perf pmu data structures */
+pthread_mutex_t fpga_perf_lock = PTHREAD_RECURSIVE_MUTEX_INITIALIZER_NP;
```

Comment:
Since we are no longer tracking a global variable, but are instead passing an fpga_perf_counter * to the API functions, we no longer need this global mutex (fpga_perf_lock).  Instead, let's add a pthread_mutex_t to struct fpga_perf_counter and lock/unlock it as needed.